src/software/thunderscope/thunderscope.py

```diff
@@ -47,46 +74,227 @@ class Thunderscope(object):
+            shelf["yellow_dock_state"] = self.yellow_full_system_dock_area.saveState()
+
+        with shelve.open(SAVED_LAYOUT_PATH, "c") as shelf:
+            shelf["blue_dock_state"] = self.blue_full_system_dock_area.saveState()
+            shelf["yellow_dock_state"] = self.yellow_full_system_dock_area.saveState()
+
+    def load_layout(self, filename=None):
+        """Open a file dialog to load the layout and state to all widgets
+
+        :param filename: The filename to load the layout from. If None, then
+                         open a file dialog.
+
+        """
+
+        if filename is None:
+            filename, _ = QtGui.QFileDialog.getOpenFileName(
+                self.window,
+                "Open layout",
+                "~/",
+                options=QFileDialog.Option.DontUseNativeDialog,
+            )
+
+            if not filename:
+                print("No filename selected")
+                return
+
+        with shelve.open(filename, "r") as shelf:
+            self.blue_full_system_dock_area.restoreState(
+                shelf["blue_dock_state"], missing="ignore"
+            )
+            self.yellow_full_system_dock_area.restoreState(
+                shelf["yellow_dock_state"], missing="ignore"
+            )
+
+            # Update default layout
+            if filename != SAVED_LAYOUT_PATH:
+                with shelve.open(SAVED_LAYOUT_PATH, "c") as default_shelf:
+                    default_shelf["blue_dock_state"] = shelf["blue_dock_state"]
+                    default_shelf["yellow_dock_state"] = shelf["yellow_dock_state"]
+                    default_shelf.sync()
+
+    def load_saved_layout(self, layout_path, load_blue=True, load_yellow=True):
+        """Load the specified layout or the default file. If the default layout
+        file doesn't exist, and no layout is provided, then just configure
+        the default layout.
+
+        :param layout_path: Path to the layout file to load.
+        :param load_blue: Whether to load the blue layout.
+        :param load_yellow: Whether to load the yellow layout.
+
+        """
+        if load_yellow:
+            self.configure_default_layout(
+                self.yellow_full_system_dock_area,
+                self.simulator_proto_unix_io,
+                self.yellow_full_system_proto_unix_io,
+                True,
+            )
+
+        if load_blue:
+            self.configure_default_layout(
+                self.blue_full_system_dock_area,
+                self.simulator_proto_unix_io,
+                self.blue_full_system_proto_unix_io,
+                False,
+            )
+
+        path = layout_path if layout_path else SAVED_LAYOUT_PATH
+
+        try:
+            self.load_layout(path)
+
+        except Exception as exc:
+            print(
+                exc,
+                Warning(
+                    "No layout file specified and default "
+                    + "layout at {} doesn't exist".format(path)
```

Comment:
doesn't seem like this is being create in setup_software?

Reply:
no it creates a layout when you hit cntrl-s, you can do that even with the default layout.

I don't like handling an exception with a pass and not logging anything which is why I had this print here. But since it seems to be an issue, I've removed it. 

It won't log this msg anymore if you chose to use the default layout.